os/drivers/bluetooth/virtual_ble/ble_vdev.c

```diff
@@ -19,8 +19,873 @@
+    if (res < 0) {
+        VBLE_ERROR(res);
+    }
+}
+
+static inline void _send_signal(struct vble_msg *msg)
+{
+    int res = sem_post(msg->signal);
+    if (res < 0) {
+        VBLE_ERROR(res);
+    }
+}
+
+static inline int _recv_message(int fd, char *buf, int buflen)
+{
+    int received = 0;
+    while (1) {
+        int res = read(fd, (void *)(buf + received), buflen - received);
+        if (res < 0) {
+            VBLE_ERROR(res);
+            return -1;
+        }
+        received += res;
+        if (received == buflen) {
+            break;
+        }
+    }
+    return 0;
+}
+
+static inline int _send_message(int fd, char *buf, int buflen)
+{
+    int sent = 0;
+    while (1) {
+        int res = write(fd, (void *)(buf + sent), buflen - sent);
+        if (res < 0) {
+            VBLE_ERROR(res);
+            return -1;
+        }
+        sent += res;
+        if (sent == buflen) {
+            break;
+        }
+    }
+    return 0;
+}
+
+static inline int _progress_message(struct vble_req *req)
+{
+    int fd = open(VBLE_MSG_QUEUE_NAME, O_WRONLY);
+    if (fd < 0) {
+        VBLE_ERROR(0);
+        return -1;
+    }
+
+    struct vble_msg *msg = NULL;
+    int res = _create_message(&msg, req);
+    if (res < 0) {
+        VBLE_ERROR(res);
+        close(fd);
+        return -1;
+    }
+
+    res = _send_message(fd, (char *)msg, sizeof(struct vble_msg));
+    close(fd);
+    if (res < 0) {
+        VBLE_ERROR(res);
+        _destroy_message(msg);
+        return -1;
+    }
+
+    _wait_message(msg);
+    _destroy_message(msg);
+
+    return 0;
+}
+
+int _vble_create_msgqueue(int *fd)
+{
+    int res = mkfifo(VBLE_MSG_QUEUE_NAME, 0666);
+    if (res < 0 && res != -EEXIST) {
+        VBLE_ERROR(0);
+        return -1;
+    }
+
+    *fd = open(VBLE_MSG_QUEUE_NAME, O_RDWR);
+    if (*fd < 0) {
+        VBLE_ERROR(0);
+        unlink(VBLE_MSG_QUEUE_NAME);
+        return -1;
+    }
+
+    return 0;
+}
+
+
+/*
+ * Handler
+ */
+static void vdev_run(int argc, char *argv[])
+{
+    int fd;
+    int res = _vble_create_msgqueue(&fd);
+    if (res < 0) {
+        VBLE_ERROR(res);
+        return;
+    }
+
+    fd_set rfds, tfds;
+    FD_ZERO(&rfds);
+    FD_SET(fd, &rfds);
+
+    while (1) {
+        tfds = rfds;
+        res = select(fd + 1, &tfds, NULL, NULL, NULL);
+        if (res <= 0) {
+            if (errno == EINTR) {
+                continue;
+            }
+            VBLE_ERROR(res);
+            break;
+        }
+        struct vble_msg msg;
+        res = _recv_message(fd, (char *)&msg, sizeof(struct vble_msg));
+        if (res < 0) {
+            VBLE_ERROR(res);
+            break;
+        }
+
+        res = vble_handle_message(msg.req);
+        if (res < 0) {
+            VBLE_ERROR(res);
+            break;
+        }
+        _send_signal(&msg);
+    }
+    return;
+}
+
+void vble_start(void){
+    g_vble_dev = bledev_register(&g_trble_drv_ops);
+    if (!g_vble_dev) {
+        VBLE_ERROR(0);
+        return;
```

Comment:
This return and line 331 are unnecessary.

Reply:
Done